src/redux/slices/userSlice.js

```diff
@@ -0,0 +1,124 @@
+
+import {
+    getUser, setUserInLocalStorage, addToFollowing,
+    addToFollowers, removeFromFollowers, removeFromFollowing, postProfileData
+} from "./userFunctions";
+
+import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
+import { getAuth, createUserWithEmailAndPassword, signInWithEmailAndPassword, signOut } from "@firebase/auth";
+import { getDocs, addDoc, arrayRemove } from "@firebase/firestore";
+import { usersRef } from "../../firebase";
+
+//logged in user
+const initialState = localStorage.getItem("user") === null ? {} : JSON.parse(localStorage.getItem("user"));
+
+const auth = getAuth();
+
+
+const addUserToDB = async (fullName, username, uid) => {
+    try {
+        const res = await addDoc(usersRef, {
+            ...initialState, fullName: fullName, username: username, uid: uid
+        });
+        setUserInLocalStorage(uid);
+    }
+    catch (e) {
+        console.log(e);
+    }
+}
+
+export const userSignUp = createAsyncThunk(
+    "user/signup",
+    async (arg) => {
+        try {
+            const res = await createUserWithEmailAndPassword(auth, arg.email, arg.password);
+            addUserToDB(arg.fullName, arg.username, res.user.uid)
+            return res.user;
+        } catch (e) { console.log(e) }
+    }
+)
+
+export const userLogin = createAsyncThunk(
+    "user/login",
+    async (arg) => {
+        try {
+            const res = await signInWithEmailAndPassword(auth, arg.email, arg.password);
+            setUserInLocalStorage(res.user.uid);
+            const loggedUser = getUser(res.user.uid);
+            return loggedUser;
+        }
+        catch (e) {
+            console.log(e);
+            return e;
+        }
+    }
+)
+
+export const userLogout = createAsyncThunk("user/logout", async (arg) => {
+    try {
+        const res = await signOut(auth);
+        localStorage.removeItem("user");
+        return null;
+    }
+    catch (e) {
+        console.log(e);
+        return false;
+    }
+})
+
+export const userSlice = createSlice({
+    name: "user",
+    initialState,
+    reducers: {},
+    extraReducers: {
+        [userSignUp.fulfilled]: (state, action) => {
+            state.email = action.payload.email
+        },
+        [userLogin.fulfilled]: (state, action) => {
+            state.uid = action.payload.uid;
+            state.fullName = action.payload.fullName;
+            state.username = action.payload.username;
+            state.posts = action.payload.posts;
+            state.bookmarks = action.payload.bookmarks;
+            state.likes = action.payload.likes;
+            state.followers = action.payload.followers;
+            state.following = action.payload.following;
+            state.bio = action.payload.bio;
+            state.portfolioURL = action.payload.portfolioURL;
+            state.uid = action.payload.uid;
+            state.id = action.payload.id;
+            state.profilepic = action.payload.profilepic;
```

Comment:
Instead of writing so much have you tried writing:
`state=action.payload ` As we are writing mutable logic in the Redux reducers so there won't be any worries about data not matching. If it doesn't work then you can keep using the same format as earlier

Reply:
Done, this is working for the case where page is redirected, so implemented here 👍🏼